hamt/hamt.go

```diff
@@ -355,7 +352,29 @@ func (ds *Shard) EnumLinksAsync(ctx context.Context) <-chan format.LinkResult {
 		defer cancel()
 		getLinks := makeAsyncTrieGetLinks(ds.dserv, linkResults)
 		cset := cid.NewSet()
-		err := dag.Walk(ctx, getLinks, ds.cid, cset.Visit, dag.Concurrent())
+		// FIXME: The `dag.Walk` interface depends on the CID of the root of
+		//  the DAG being traversed.
+		//  When this function is called from `io.sizeBelowThreshold` we may
+		//  not have that available (as the directory has changed and the last
+		//  CID from which the `io.HAMTDirectory` was loaded from, if any, might
+		//  be outdated).
+		//  Ideally we would need to extend the DAG walk function to handle Shards
+		//  and not IPLD nodes. (See https://github.com/ipfs/go-ipld-format/blob/master/walker.go
+		//  as an example of a more generic walker that doesn't necessarily depend
+		//  on IPLD nodes and CIDs, although it may not be the correct choice here.)
+		//
+		//  For the moment, to make the current PR tests pass, we force the Shards
+		//  to serialize to an IPLD node to extract its CID for the `Walk` API,
+		//  but this of course means fetching all the shards in the directory
+		//  beforehand which is exactly what we don't want (defeating the initial
+		//  purpose of the `io.sizeBelowThreshold` that tries to retrieve as little
+		//  shards as possible).
+		rootNode, err := ds.Node()
+		if err != nil {
+			emitResult(ctx, linkResults, format.LinkResult{Link: nil, Err: err})
+			return
+		}
+		err = dag.Walk(ctx, getLinks, rootNode.Cid(), cset.Visit)
```

Comment:
Blocking.

Reply:
I'm slightly confused is this an existing bug? If `EnumLinksAsync` takes a shard CID, but the shard objects are mutable and don't update their CIDs shouldn't we already be susceptible to problems here?

Reply:
Yes, I think this is an existing issue (for the `HAMTDirectory`, not for the `BasicDirectory`). As mentioned before I can submit a new issue to discuss this independently if it works for you.

My guess is that this wasn't spotted before because we didn't rely so much on the internals of the `EnumLinksAsync` logic as we do in the sharding case.

Reply:
Ok, this seems fine then. Let's also just delete the `Shard.cid` variable since we're no longer using it.